Add WideString and WideText.

Xitian9 opened the pull request:
These calculate width differently than string and text, taking into account that some characters are wider than 1 space (e.g. Han characters) and some are narrower (e.g. unicode combining characters).

Also adds some instances for Text and Text Builder.

Fixes #8.

Xitian9 committed:
Add WideString.

This is a newtype wrapper for String which calculates width based on the
monospace width of characters. For example, Han ideographs are 2 wide,
and unicode combining characters are 0 wide.

Xitian9 committed:
Include instances and types for Text.

Since we now depend on doclayout, our transitive dependencies include
text, so we aren't made any lighter weight by omitting these.

This adds Cell and StringBuilder instances for Text, and a StringBuilder
instance for a Text Builder. It also adds WideText, a newtype wrapper
around Text akin to WideString.

muesli4 commented:
I will probably need more time to review this.

* Originally, I wanted to keep the dependencies simple.  If `text` is really widespread and installed everywhere, then I don't see any issues with incorporating this.
* Do you think `WideString` and `WideText` could be unified with a type class?
* I'm not sure if I want the dependency to `doclayout` but I haven't made up my mind yet.  Another package `table-layout-wide` would be an alternative.

Xitian9 commented:
I think `text` is widespread enough that if you're writing a text handling-library, you're pretty safe incorporating it.


By unifying with a type class, I presume you mean:
```
newtype WideString a = WideString a

instance Cell (WideString String) where …
instance Cell (WideString Text) where…
```

I considered unifying them in a type class. I would be happy to do that, but I initially decided against it for a couple of reasons:

1. There are really only three types that we would want instances for: `String`, `Text`, and lazy `Text`
2. It would help type inference and be shorter to refer to `WideString` and `WideText` instead of `WideString String` and `WideString Text`.
3. The most general definition would be `instance HasChars a => Cell (WideString a)`, but the typeclass methods actually wouldn't let you short-circuit within `dropLeft`, so you would probably want separate definitions for each anyway.


If you want to break out the wide stuff into a separate package I would be happy with that too. The only transitive dependencies that `doclayout` adds are `doclayout` itself and `safe`.

muesli4 commented on src/Text/Layout/Table/Cell/WideString.hs:
```diff
@@ -0,0 +1,77 @@
+
+-- | A newtype for String in which characters can be wider than one space.
+newtype WideString = WideString String
+    deriving (Eq, Ord, Show, Read, Semigroup, Monoid, IsString, StringBuilder)
```
What would be the advantage of a `StringBuilder` instance?

Xitian9 replied:
Probably nothing. There's a `StringBuilder` instance for `String` (and now `Text` as well), so I added these for symmetry. I'll remove.

muesli4 commented on src/Text/Layout/Table/Cell/WideString.hs:
```diff
@@ -0,0 +1,77 @@
+
+-- | A newtype for Text in which characters can be wider than one space.
+newtype WideText = WideText T.Text
+    deriving (Eq, Ord, Show, Read, Semigroup, Monoid, IsString, StringBuilder)
```
Same here about `StringBuilder`.

muesli4 commented on src/Text/Layout/Table/Cell/WideString.hs:
```diff
@@ -0,0 +1,77 @@
+dropWide _ i [] = []
+dropWide gobbleZeroWidth i (x:xs)
+    | gobbleZeroWidth && i == 0 && l == 0 = dropWide gobbleZeroWidth i xs
+    | i <= 0    = x:xs
```
Please use spaces around operators and also in patterns.  Actually, I would prefer if you could use a named pattern and use it here instead of constructing a new list.
```
dropWide gobbleZeroWidth i l@(x : xs)
...
    | i <= 0     = l
```

muesli4 commented on src/Text/Layout/Table/Cell/WideString.hs:
```diff
@@ -0,0 +1,77 @@
+
+    buildCell (WideString s) = buildCell s
+
+dropWide :: Bool -> Int -> String -> String
```
Even though it is non-public, please provide at least a sentence of Haskell documentation.

muesli4 commented on src/Text/Layout/Table/Cell/WideString.hs:
```diff
@@ -0,0 +1,77 @@
+    | i <= 0    = x:xs
+    | l <= i    = dropWide gobbleZeroWidth (i - l) xs
+    | otherwise = replicate (l - i) ' ' ++ xs
+  where l = charWidth x
```
Please add a newline for definitions with 2 spaces indentation.

muesli4 replied:
`l` seems ambiguous here.  Perhaps, use `charLen`.

muesli4 commented on src/Text/Layout/Table/Cell/WideString.hs:
```diff
@@ -0,0 +1,77 @@
+dropLeftWideT :: Int -> T.Text -> T.Text
+dropLeftWideT i txt = case T.uncons txt of
+    Nothing -> txt
+    Just (x,xs) -> let l = charWidth x in if
+        | i == 0 && l == 0 -> dropLeftWideT i xs
+        | i <= 0    -> txt
+        | l <= i    -> dropLeftWideT (i - l) xs
+        | otherwise -> T.replicate (l - i) " " <> xs
+
+dropRightWideT :: Int -> T.Text -> T.Text
+dropRightWideT i txt = case T.unsnoc txt of
+    Nothing -> txt
+    Just (xs,x) -> let l = charWidth x in if
```
Please use space after comma in tuple pattern.

muesli4 commented on src/Text/Layout/Table/Cell/WideString.hs:
```diff
@@ -0,0 +1,77 @@
+
+    buildCell (WideText s) = buildCell s
+
+dropLeftWideT :: Int -> T.Text -> T.Text
+dropLeftWideT i txt = case T.uncons txt of
+    Nothing -> txt
+    Just (x,xs) -> let l = charWidth x in if
+        | i == 0 && l == 0 -> dropLeftWideT i xs
+        | i <= 0    -> txt
+        | l <= i    -> dropLeftWideT (i - l) xs
+        | otherwise -> T.replicate (l - i) " " <> xs
+
+dropRightWideT :: Int -> T.Text -> T.Text
+dropRightWideT i txt = case T.unsnoc txt of
+    Nothing -> txt
+    Just (xs,x) -> let l = charWidth x in if
+        | i <= 0    -> txt
+        | l <= i    -> dropRightWideT (i - l) xs
+        | otherwise -> xs <> T.replicate (l - i) " "
```
These are identical except for the `otherwise` case.  Can you think of some way to refactor that?

Xitian9 replied:
I'll try to think of a way. They're actually more different than they look: note that one uses `uncons` and the other uses `unsnoc`. This means their pattern matches are reversed too (one is `(x,xs)`, the other is `(xs,x)`).

muesli4 replied:
@Xitian9, oh, my bad.  Then just forget about it. :)

muesli4 commented on test-suite/TestSpec.hs:
```diff
@@ -171,6 +175,51 @@ spec = do
+            prop "agrees for ascii strings" $ \(ASCIIString x) -> buildCell (WideString x) `shouldBe` x
+            it "renders double width" $ buildCell wide `shouldBe` "㐀㐁㐂"
+            it "renders zero width" $ buildCell narrow `shouldBe` "Bien sûr!"
+        describe "visibleLength" $ do
+            prop "agrees for ascii strings" $ \(ASCIIString x) -> visibleLength (WideString x) `shouldBe` visibleLength x
+            it "detects double width" $ visibleLength wide `shouldBe` 6
+            it "detects zero width" $ visibleLength narrow `shouldBe` 9
+        describe "measureAlignment" $ do
+            prop "agrees for ascii strings" $ \(ASCIIString x) -> measureAlignment (=='e') (WideString x) `shouldBe` measureAlignment (=='e') x
+            it "detects double width" $ measureAlignment (=='㐁') wide `shouldBe` AlignInfo 2 (Just 2)
+            it "fails to detect" $ measureAlignment (=='a') wide `shouldBe` AlignInfo 6 Nothing
+            it "detects zero width after" $ measureAlignment (=='n') narrow `shouldBe` AlignInfo 3 (Just 5)
+            it "detects zero width before" $ measureAlignment (=='r') narrow `shouldBe` AlignInfo 7 (Just 1)
+        describe "dropLeft" $ do
+            prop "agrees for ascii strings" $ \(Small n) (ASCIIString x) -> buildCell (dropLeft n (WideString x)) `shouldBe` dropLeft n x
+            describe "on wide characters" $ do
+                it "drops 1 character of double width" $ dropLeft 2 wide `shouldBe` WideString "㐁㐂"
+                it "drops 2 characters of double width and adds a space" $ dropLeft 3 wide `shouldBe` WideString " 㐂"
+            describe "on narrow characters" $ do
+                it "drops combining characters with their previous" $ dropLeft 7 narrow `shouldBe` WideString "r!"
+        describe "dropRight" $ do
+            prop "agrees for ascii strings" $ \(Small n) (ASCIIString x) -> buildCell (dropRight n (WideString x)) `shouldBe` dropRight n x
+            describe "on wide characters" $ do
+                it "drops 1 character of double width" $ dropRight 2 wide `shouldBe` WideString "㐀㐁"
+                it "drops 2 characters of double width and adds a space" $ dropRight 3 wide `shouldBe` WideString "㐀 "
+            describe "on narrow characters" $ do
+                it "drops a combining character for free" $ dropRight 3 narrow `shouldBe` WideString "Bien s"
+                it "does not drop a combining character without their previous" $ dropRight 2 narrow `shouldBe` WideString "Bien sû"
+
+    describe "wide text" $ do
+        describe "buildCell" $ do
+            prop "gives the same result as wide string" $ \x -> buildCell (WideText $ T.pack x) `shouldBe` x
+        describe "visibleLength" $ do
+            prop "gives the same result as wide string" $ \x -> visibleLength (WideText $ T.pack x) `shouldBe` visibleLength (WideString x)
+        describe "measureAlignment" $ do
+            prop "gives the same result as wide string" $ \x -> measureAlignment (=='e') (WideText $ T.pack x) `shouldBe` measureAlignment (=='e') (WideString x)
+        describe "dropLeft" $ do
+            prop "gives the same result as wide string" $ \(Small n) x -> buildCell (dropLeft n . WideText $ T.pack x) `shouldBe` dropLeft n (WideString x)
+        describe "dropRight" $ do
+            prop "gives the same result as wide string" $ \(Small n) x -> buildCell (dropRight n . WideText $ T.pack x) `shouldBe` dropRight n (WideString x)
```
Please use space after `==`.  This also looks like a pattern that should be abstracted:
```
    measureAlignmentAt c = measureAlignment (== c)
```

muesli4 reviewed (commented):
I agree with the reasoning you provided and will accept adding those dependencies.

Xitian9 commented:
I've made the requested changes (I think).

I also corrected a bug and added a test. The bug occurred where, when dropping from the left, zero-width characters which immediately followed characters of width 2 or greater were not dropped when the wide character would have been cut in half (and so dropped and padded with space).

To demonstrate, let XX be a character of width 2, 0 a character of width 0, and Y another character.
`dropLeft 1 "XX0Y"` should be `" Y"`, but previously it would have been `" 0Y"`.

muesli4 commented:
I resolved the conflicts locally and rebased manually.

elldritch commented:
I'm really looking forward to seeing this release! I've been working on a quick project that displays Twitter profile names, and some profile names with emojis or other Unicode seem to cause misalignment. In case you want to test whether this fix addresses that, here are some examples:

```
+------+--------------+-----------------+---------------------------------------------------+-------------------------+-----------+-----------+-----------+--------+
| 14   | 36           | rakyll          | Jaana Dogan ヤナ ドガン                                | 2007-12-03 13:08:02 UTC | True      | 73260     | 1084      | 86769  |
+------+--------------+-----------------+---------------------------------------------------+-------------------------+-----------+-----------+-----------+--------+
```

```
+------+--------------+-----------------+---------------------------------------------------+-------------------------+-----------+-----------+-----------+--------+
| 32   | 20           | ashleyloob      | 🍍ashley🍍                                          | 2020-01-17 07:07:17 UTC | False     | 21505     | 400       | 8536   |
+------+--------------+-----------------+---------------------------------------------------+-------------------------+-----------+-----------+-----------+--------+
```

```
+------+--------------+-----------------+---------------------------------------------------+-------------------------+-----------+-----------+-----------+--------+
| 52   | 13           | ryantakesoff    | Ryan Breslow 🕺                                    | 2011-09-07 21:18:44 UTC | True      | 25770     | 525       | 3669   |
+------+--------------+-----------------+---------------------------------------------------+-------------------------+-----------+-----------+-----------+--------+
```

Xitian9 commented:
It looks like your examples don't quite align properly. I'll do some more testing to see if the bug can be fixed here or if it's an upstream issue with `doclayout`.

Xitian9 commented:
I get the katakana aligning properly on my screen, but the emoji are incorrectly detected as a single space wide, leading to misalignment. I think this is an upstream bug. (Reported [here](https://github.com/jgm/doclayout/issues/3))
